scripts/gen_visitors.py

```diff
@@ -0,0 +1,293 @@
+                    ):
+                        child_desc = key_fd.message_type
+                        child_needed = self.walk(child_desc)
+                        if child_needed:
+                            needed = True
+                            lines.append(
+                                emit_loop(
+                                    field.name,
+                                    f"o.{field.name}.keys()",
+                                    name_for(child_desc),
+                                )
+                            )
+                else:
+                    child = self.check_repeated(
+                        field.message_type, field, f"o.{field.name}"
+                    )
+                    if child is not None:
+                        needed = True
+                        lines.append(child)
+            else:
+                child_desc = field.message_type
+                child_needed = self.walk(child_desc)
+                needed |= child_needed
+                if child_needed:
+                    lines.append(
+                        emit_singular(
+                            field.name, f"o.{field.name}", name_for(child_desc), True
+                        )
+                    )
+
+        # Process oneof fields as if/elif chains
+        for oneof_idx, fields in oneof_fields.items():
+            oneof_lines = []
+            first = True
+            for field in fields:
+                child_desc = field.message_type
+                child_needed = self.walk(child_desc)
+                needed |= child_needed
+                if child_needed:
+                    if_word = "if" if first else "elif"
+                    first = False
+                    line = emit_singular(
+                        field.name, f"o.{field.name}", name_for(child_desc), True
+                    ).replace("        if", f"        {if_word}", 1)
+                    oneof_lines.append(line)
+            if oneof_lines:
+                lines.extend(oneof_lines)
+
+        self.generated[key] = needed
+        self.in_progress.discard(key)
+        if needed:
+            self.methods.append("\n".join(lines) + "\n")
+        return needed
+
+    def generate(self, roots: list[Descriptor]) -> str:
+        """
+        Generate Python source code that, given a function f(Payload) -> Payload,
+        applies it to every Payload contained within a WorkflowActivation tree.
+
+        The generated code defines async visitor functions for each reachable
+        protobuf message type starting from WorkflowActivation, including support
+        for repeated fields and map entries, and a convenience entrypoint
+        function `visit`.
+        """
+
+        # We avoid importing google.api deps in service protos; expand by walking from
+        # WorkflowActivationCompletion root which references many command messages.
+        for r in roots:
+            self.walk(r)
```

Comment:
Just a readability suggestion but I'd move this method to the top of the class since it has the nice docstring and the header introduces the reader to the problem of generating  correctly indented methods that the rest of the class's methods address.